src/core/api/api_access_control.pl

```diff
@@ -0,0 +1,486 @@
+              api_delete_role/3,
+              api_get_organizations/3,
+              api_get_organization_from_name/4,
+              api_get_organization_from_id/4,
+              api_add_organization/4,
+              api_delete_organization/3,
+              api_grant_capability/3,
+              api_revoke_capability/3,
+              api_get_user_from_name/4,
+              api_get_resource_from_name/4,
+              api_add_user/4,
+              api_delete_user/3,
+              api_get_users/3,
+              api_get_user_from_name/4,
+              api_get_user_from_id/4,
+              api_update_user_password/4
+          ]).
+
+:- use_module(core(util)).
+:- use_module(core(account)).
+:- use_module(core(query)).
+:- use_module(core(document)).
+:- use_module(core(transaction)).
+:- use_module(library(plunit)).
+:- use_module(library(crypto)).
+:- use_module(library(lists)).
+:- use_module(library(apply)).
+:- use_module(library(yall)).
+
+api_get_roles(SystemDB, Auth, Roles) :-
+    do_or_die(
+        is_super_user(Auth),
+        error(access_not_authorised(Auth,'Action/manage_capabilities','SystemDatabase'), _)),
+    get_roles(SystemDB, Roles).
+
+get_roles(SystemDB,Roles) :-
+    findall(Role,
+            ask(SystemDB,
+                (   t(Id, rdf:type, '@schema':'Role'),
+                    get_document(Id, Role)
+                )),
+            Roles).
+
+api_get_role_from_id(SystemDB, Auth, Id, Role) :-
+    do_or_die(
+        is_super_user(Auth),
+        error(access_not_authorised(Auth,'Action/manage_capabilities','SystemDatabase'), _)),
+    do_or_die(
+        get_role_from_id(SystemDB, Id, Role),
+        error(no_role_with_given_id(Id), _)
+    ).
+
+%% get_role_from_name(+,+,-) is det + error
+get_role_from_id(SystemDB, Id, Role) :-
+    ask(SystemDB,
+        (   t(Id, rdf:type, '@schema':'Role'),
+            get_document(Id, Role)
+        )).
+
+api_get_role_from_name(SystemDB, Auth, RoleName, Role) :-
+    do_or_die(
+        is_super_user(Auth),
+        error(access_not_authorised(Auth,'Action/manage_capabilities','SystemDatabase'), _)),
+    get_role_from_name(SystemDB, RoleName, Role).
+
+%% get_role_from_name(+,+,-) is det + error
+get_role_from_name(SystemDB, RoleName, Role) :-
+    findall(Role,
+            ask(SystemDB,
+                (   t(Id, name, RoleName^^xsd:string),
+                    t(Id, rdf:type, '@schema':'Role'),
+                    get_document(Id, Role)
+                )),
+            Roles),
+    (   Roles = [Role]
+    ->  true
+    ;   Roles = []
+    ->  throw(error(no_id_for_role_name(RoleName), _))
+    ;   throw(error(no_unique_id_for_role_name(RoleName),_))
+    ).
+
+api_add_role(_, Auth, Role, Id) :-
+    do_or_die(
+        is_super_user(Auth),
+        error(access_not_authorised(Auth,'Action/manage_capabilities','SystemDatabase'), _)),
+
+    % Make sure that we only insert role types.
+    put_dict('@type', Role, 'Role', Typed_Role),
```

Comment:
I don't think we should just insert the Role document (with added type), due to concerns around data validation. As a general principle I think we should be very very suspicious of any incoming user data, especially if these are records that have to go into the system database. Instead of just inserting them directly (after slight modifications), I'm in favor of always building these sorts of documents ourselves, filling out details using the input document. This ensures we take a whitelist approach to what goes into our database, rather than just trusting the user to not submit documents with extra unexpected fields, or blacklisting particular fields we don't trust.